sql/core/src/main/scala/org/apache/spark/sql/ArrowConverters.scala

```diff
@@ -83,53 +83,41 @@ private[sql] object ArrowConverters {
   }
 
   /**
-   * Transfer an array of InternalRow to an ArrowRecordBatch.
+   * Iterate over InternalRows and write to an ArrowRecordBatch.
    */
-  private[sql] def internalRowsToArrowRecordBatch(
-      rows: Array[InternalRow],
+  private def internalRowIterToArrowBatch(
+      rowIter: Iterator[InternalRow],
       schema: StructType,
-      allocator: RootAllocator): ArrowRecordBatch = {
-    val fieldAndBuf = schema.fields.zipWithIndex.map { case (field, ordinal) =>
-      internalRowToArrowBuf(rows, ordinal, field, allocator)
+      allocator: RootAllocator,
+      initialSize: Option[Int]): ArrowRecordBatch = {
+
+    val columnWriters = schema.fields.zipWithIndex.map { case (field, ordinal) =>
+      ColumnWriter(ordinal, allocator, field.dataType)
+        .init(initialSize)
+    }
+
+    rowIter.foreach { row =>
```

Comment:
Yeah, we even discussed this before I think..  Even here it might add a little overhead to wrap in a function object, so I'll change it to a while loop just to be sure.